### PR TITLE
Introduce OnchainTxHandler, move bumping and tracking logic

### PR DESCRIPTION
**Pull request by ariard**
Build on top of #461, unify transaction generation code in its own nodule.

Simplify parsing code (easier to implement option_simplified_commitment) and first step towards removal of in-memory private keys from ChannelMonitor.

Note: there is one intermediary failure on e2a3bd6, which is related to #459, need to rebase also on top of it when it's merged.

### CODE REVIEW

File: lightning/src/ln/onchaintx.rs [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+	}
+
+	pub(super) fn get_witnesses_weight(inputs: &[InputDescriptors]) -> usize {
+		let mut tx_weight = 2; // count segwit flags
+		for inp in inputs {
+			// We use expected weight (and not actual) as signatures and time lock delays may vary
+			tx_weight +=  match inp {
+				// number_of_witness_elements + sig_length + revocation_sig + pubkey_length + revocationpubkey + witness_script_length + witness_script
+				&InputDescriptors::RevokedOfferedHTLC => {
+					1 + 1 + 73 + 1 + 33 + 1 + 133
```
Comment (devrandom) [inline lightning/src/ln/onchaintx.rs]: not for this PR, but I wonder if it will be more robust to create a sample tx and serialize it to get these numbers

Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Sure further dry-up of this code with some tx template what somehow in my mind

File: lightning/src/ln/onchaintx.rs [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+					let new_fee = if $old_feerate < $fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::HighPriority) {
+						let mut value = $amount;
+						if subtract_high_prio_fee!(self, $fee_estimator, value, $predicted_weight, used_feerate) {
+							// Overflow check is done in subtract_high_prio_fee
+							$amount - value
+						} else {
+							log_trace!(self, "Can't new-estimation bump new claiming tx, amount {} is too small", $amount);
+							return None;
+						}
+					// ...else just increase the previous feerate by 25% (because that's a nice number)
```
Comment (devrandom) [inline lightning/src/ln/onchaintx.rs]: (I know this is not changed in this PR)

Might want to consider escalating the fee much faster, in case the fee estimator is wrong.  Even switching to a doubling mode after some time.  The game theory is that we might lose the entire channel value, so even if we end up paying 10% in fees it may be worth it.

Also, if the fee estimator increases moderately for a long time, this branch will never be taken, and again if the fee estimator is wrong we can lose funds.

File: lightning/src/ln/channelmonitor.rs [hunk 3]
```diff
@@ -1766,319 +1445,83 @@ impl ChannelMonitor {
...
-										sequence: idx as u32, // reset to 0xfffffffd in sign_input
-										witness: Vec::new(),
-									};
-									if htlc.cltv_expiry > height + CLTV_SHARED_CLAIM_BUFFER {
-										inputs.push(input);
-										inputs_desc.push(if htlc.offered { InputDescriptors::OfferedHTLC } else { InputDescriptors::ReceivedHTLC });
-										inputs_info.push((payment_preimage, tx.output[transaction_output_index as usize].value, htlc.cltv_expiry));
-										total_value += tx.output[transaction_output_index as usize].value;
-									} else {
-										let mut single_htlc_tx = Transaction {
```
Comment (devrandom) [inline lightning/src/ln/channelmonitor.rs]: I'm probably missing something, but I don't see where this code moved.

**Comment by TheBlueMatt** [comment]
Was kinda hoping #461 went in first, but waiting to take a look at this for concept acks before that was unfair. Sorry about that.

At a high-level design-wise, I like the direction, though I'm not sure *exactly* what the end-goal is. I presume the eventual goal is to just have ChannelMonitor watch the chain with no way to sign anything, and then just call OnchainTxHandler to get the signed transactions. Thus, the monitoring would be disconnected from the signer, and OnchainTxHandler could contain only pre-signed transactions. However, that leaves me confused why the RBF bumping would be in OnchainTxHandler.

Alternatively, OnchainTxHandler could be responsible for *monitoring* any outputs which are "to-be-claimed", with ChannelMonitor solely responsible for detecting such outputs and notifying ChannelManager appropriately after OnchainTxHandler has done its job, though that would leave OnchainTxHandler with a *further* substruct which could either be presigned txn or with a ChannelKeys/Signer.

**Comment by ariard** [comment]
> Alternatively, OnchainTxHandler could be responsible for monitoring any outputs which are "to-be-claimed", with ChannelMonitor solely responsible for detecting such outputs and notifying ChannelManager appropriately after OnchainTxHandler has done its job, though that would leave OnchainTxHandler with a further substruct which could either be presigned txn or with a ChannelKeys/Signer.

End-goals are unifying transaction generation code, encapsulate in-flight and bumping logic complexity in its own module and simplify tx parsing code (a.k.a the detection). Like you said, next step is obviously to add Signer in OnchainTxHandler. Signer implementation would determine if it's in watchtower mode or local mode (sign_justice_tx would return pre-signed txn for watchtower). 

ChannelMonitor would be left with only channel tx detection and passing preimage/timeout to offchain, which are already complex tasks (specially if we have different commitment tx format like first option_simplified_commitment).

Long-term, OnchainTxHandler could be reused for dual-funding tx/splicing, where we may have to bump onchain txn too (I don't want to write another bumping logic).

> However, that leaves me confused why the RBF bumping would be in OnchainTxHandler.

I still need to think about this but IMO _what_ you can sign determine _how_ you can bump (no key access for watchtower means CPFP or justice tx instead of RBF).

**Comment by TheBlueMatt** [comment]
Right, detection-vs-response is a good distinction, I just wanted to make sure that was the goal, and note that its not the only thing we need.

File: lightning/src/ln/onchaintx.rs (hunk 2 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: I agree picking up a %25 bump was just magic and more aggressive heuristics could be implemented. So far IIRC we are the only implem bumping fees would like more feedback from other LN devs.

The game theory remark is a very good one.

File: lightning/src/ln/channelmonitor.rs (hunk 3 above)
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: It's L1321 as of 8603afb, aggregation now happens in OnchainTxHandler so there is no more a isolated vs multiple htlcs branch anymore.

File: lightning/src/ln/onchaintx.rs [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+				&InputDescriptors::RevokedOutput => {
+					1 + 1 + 73 + 1 + 1 + 1 + 77
+				},
+			};
+		}
+		tx_weight
+	}
+
+	fn get_height_timer(current_height: u32, timelock_expiration: u32) -> u32 {
+		if timelock_expiration <= current_height || timelock_expiration - current_height <= 3 {
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Thanks, corrected in its own commit as it was already there.

File: lightning/src/ln/channelmonitor.rs [hunk 5]
```diff
@@ -1395,12 +1220,13 @@ impl ChannelMonitor {
 	/// HTLC-Success/HTLC-Timeout transactions.
 	/// Return updates for HTLC pending in the channel and failed automatically by the broadcast of
 	/// revoked remote commitment tx
-	fn check_spend_remote_transaction(&mut self, tx: &Transaction, height: u32, fee_estimator: &FeeEstimator) -> (Vec<Transaction>, (Sha256dHash, Vec<TxOut>), Vec<SpendableOutputDescriptor>) {
+	fn check_spend_remote_transaction(&mut self, tx: &Transaction, height: u32) -> (HashMap<Sha256dHash, Vec<(u32, bool, BitcoinOutPoint, InputMaterial)>>, (Sha256dHash, Vec<TxOut>), Option<SpendableOutputDescriptor>) {
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: Good idea, added a ClaimRequest struct with more doc about field purpose.

**Comment by devrandom** [comment]
My comments were addressed.  (note that for some reason there are no "Resolve" buttons visible to me)

**Comment by ariard** [comment]
Rebased 0de8d5d, ready for new review.

File: lightning/src/ln/onchaintx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Reserved this one because got test failures due to unsigned int overflow..

File: lightning/src/ln/channelmonitor.rs [hunk 6]
```diff
@@ -383,38 +382,45 @@ struct LocalSignedTx {
...
-}
-
 /// When ChannelMonitor discovers an onchain outpoint being a step of a channel and that it needs
 /// to generate a tx to push channel state forward, we cache outpoint-solving tx material to build
 /// a new bumped one in case of lenghty confirmation delay
 #[derive(Clone, PartialEq)]
-enum InputMaterial {
+pub(crate) enum InputMaterial {
 	Revoked {
+		/// Witness script
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Note that the documentation-required checks don't apply to pub(crate) things. Two word comments probably indicate you should just change the variable name and move on :)

File: lightning/src/ln/onchaintx.rs [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+	destination_script: Script,
+
+	// Used to track claiming requests. If claim tx doesn't confirm before height timer expiration we need to bump
+	// it (RBF or CPFP). If an input has been part of an aggregate tx at first claim try, we need to keep it within
+	// another bumped aggregate tx to comply with RBF rules. We may have multiple claiming txn in the flight for the
+	// same set of outpoints. One of the outpoints may be spent by a transaction not issued by us. That's why at
+	// block connection we scan all inputs and if any of them is among a set of a claiming request we test for set
+	// equality between spending transaction and claim request. If true, it means transaction was one our claiming one
+	// after a security delay of 6 blocks we remove pending claim request. If false, it means transaction wasn't and
+	// we need to regenerate new claim request with reduced set of still-claimable outpoints.
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: If you're gonna fix a spelling mistake in a mostly-move, can you do it in a separate commit so that --color-moved doesn't barf?

File: lightning/src/ln/channelmonitor.rs [hunk 8]
```diff
@@ -628,89 +599,31 @@ pub struct ChannelMonitor<ChanSigner: ChannelKeys> {
...
+	onchain_tx_handler: OnchainTxHandler,
+
 	// We simply modify last_block_hash in Channel's block_connected so that serialization is
 	// consistent but hopefully the users' copy handles block_connected in a consistent way.
 	// (we do *not*, however, update them in insert_combine to ensure any local user copies keep
 	// their last_block_hash from its state and not based on updated copies that didn't run through
 	// the full block_connected).
 	pub(crate) last_block_hash: Sha256dHash,
-	secp_ctx: Secp256k1<secp256k1::All>, //TODO: dedup this a bit...
+	secp_ctx: Secp256k1<secp256k1::All>,
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Why? We should still think about deduping the secp contexts especially in a ManyChannelMonitor.

File: lightning/src/ln/onchaintx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/onchaintx.rs]: `timelock_expiration  <= current_height + 3` should work.

File: lightning/src/ln/onchaintx.rs [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+			version: 2,
+			lock_time: 0,
+			input: inputs,
+			output: vec![TxOut {
+				script_pubkey: self.destination_script.clone(),
+				value: 0
+			}],
+		};
+
+		macro_rules! RBF_bump {
```
Comment (arik-so) [inline lightning/src/ln/onchaintx.rs]: what's the reason for this being a macro?

File: lightning/src/ln/onchaintx.rs [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+		}
+
+		let new_timer = Self::get_height_timer(height, cached_claim_datas.soonest_timelock);
+		let mut inputs_witnesses_weight = 0;
+		let mut amt = 0;
+		for per_outp_material in cached_claim_datas.per_input_material.values() {
+			match per_outp_material {
+				&InputMaterial::Revoked { ref script, ref is_htlc, ref amount, .. } => {
+					inputs_witnesses_weight += Self::get_witnesses_weight(if !is_htlc { &[InputDescriptors::RevokedOutput] } else if HTLCType::scriptlen_to_htlctype(script.len()) == Some(HTLCType::OfferedHTLC) { &[InputDescriptors::RevokedOfferedHTLC] } else if HTLCType::scriptlen_to_htlctype(script.len()) == Some(HTLCType::AcceptedHTLC) { &[InputDescriptors::RevokedReceivedHTLC] } else { &[] });
+					amt += *amount;
```
Comment (arik-so) [inline lightning/src/ln/onchaintx.rs]: would strongly recommend using different names from these variables. Maybe revoked_amount and total_amount?

Comment (arik-so) [reply lightning/src/ln/onchaintx.rs]: Or current_revoked_amount

File: lightning/src/ln/onchaintx.rs [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+				},
+				&InputMaterial::RemoteHTLC { ref preimage, ref amount, .. } => {
+					inputs_witnesses_weight += Self::get_witnesses_weight(if preimage.is_some() { &[InputDescriptors::OfferedHTLC] } else { &[InputDescriptors::ReceivedHTLC] });
+					amt += *amount;
```
Comment (arik-so) [inline lightning/src/ln/onchaintx.rs]: same here

File: lightning/src/ln/onchaintx.rs [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+						new_fee + previous_fee + min_relay_fee - new_fee
+					} else {
+						new_fee
+					};
+					Some((new_fee, new_fee * 1000 / $predicted_weight))
+				}
+			}
+		}
+
+		let new_timer = Self::get_height_timer(height, cached_claim_datas.soonest_timelock);
```
Comment (arik-so) [inline lightning/src/ln/onchaintx.rs]: some explanation of what's happening here and what the variables will be used for could bee heelpful

File: lightning/src/ln/onchaintx.rs [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+					}
+				}
+			}
+		}
+		new_claims.push((aggregated_soonest, aggregated_claim));
+
+		// Generate claim transactions and track them to bump if necessary at
+		// height timer expiration (i.e in how many blocks we're going to take action).
+		for claim in new_claims {
+			let mut claim_material =  ClaimTxBumpMaterial { height_timer: 0, feerate_previous: 0, soonest_timelock: claim.0, per_input_material: claim.1.clone() };
```
Comment (arik-so) [inline lightning/src/ln/onchaintx.rs]: nit: 2 spacees

File: lightning/src/ln/onchaintx.rs (hunk 7 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: I don't remember to do a spelling fix there ? What git options did you use there?

File: lightning/src/ln/channelmonitor.rs (hunk 8 above)
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: Ooops, I tried at first to throw it entirely in OnchainTxHandler IIIRC. That's said, after moving local commitment txn construction there too, secp state should be the matter of the Signer

File: lightning/src/ln/functional_tests.rs [hunk 14]
```diff
@@ -2498,21 +2496,20 @@ fn test_htlc_on_chain_timeout() {
...
+
+		check_spends!(node_txn[2], commitment_tx[0].clone());
+		assert_eq!(node_txn[2].clone().input[0].witness.last().unwrap().len(), ACCEPTED_HTLC_SCRIPT_WEIGHT);
+
+		check_spends!(node_txn[0], chan_2.3.clone());
+		check_spends!(node_txn[1], node_txn[0].clone());
+		assert_eq!(node_txn[0].clone().input[0].witness.last().unwrap().len(), 71);
+		assert_eq!(node_txn[1].clone().input[0].witness.last().unwrap().len(), OFFERED_HTLC_SCRIPT_WEIGHT);
+
+		timeout_tx = node_txn[2].clone();
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/functional_tests.rs]: Right, bookmarked point, I prefer to defer this until the Great Functional Test cleanup

File: lightning/src/ln/onchaintx.rs (hunk 9 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: IIRC I though at first having different code paths for different transaction cases (revoked, valid-remote broadcast, ...) but in fact was able to subsume all them on one. I think it can be functionify when we implement CPFP as a bumping mechanism or implement a better bumping strategy (cf devrandom point above)

File: lightning/src/ln/onchaintx.rs (hunk 10 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Thanks, updated name with suggestions

File: lightning/src/ln/onchaintx.rs (hunk 12 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Yes add comment on assumptions there and also on top of get_height_timer, tell me if it's clear enough :)

**Comment by ariard** [comment]
Thanks @TheBlueMatt and @arik-so for reviewing, addressed your comment at e9dff10.

File: lightning/src/ln/onchaintx.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs]: git show --color-moved highlights this line as a different color.

**Comment by ariard** [comment]
Rebased on top #489 at 5aa3816

**Comment by TheBlueMatt** [comment]
Travis failure is
```
   --> lightning/src/ln/channelmonitor.rs:811:22

    |

811 |  secp_ctx: Secp256k1<secp256k1::All>, //TODO: dedup this a bit...

    |                      ^^^^^^^^^ Use of undeclared type or module `secp256k1`

error[E0433]: failed to resolve. Use of undeclared type or module `secp256k1`

   --> lightning/src/ln/onchaintx.rs:174:22

    |

174 |  secp_ctx: Secp256k1<secp256k1::All>,

    |                      ^^^^^^^^^ Use of undeclared type or module `secp256k1`
```

Can you rebase on master now that 489  landed?

**Comment by ariard** [comment]
Rebased bce1508, with Travis 1.22 failure fixed.

File: lightning/src/ln/channelmonitor.rs [hunk 15]
```diff
@@ -446,46 +446,37 @@ struct LocalSignedTx {
...
-}
-
 /// When ChannelMonitor discovers an onchain outpoint being a step of a channel and that it needs
 /// to generate a tx to push channel state forward, we cache outpoint-solving tx material to build
 /// a new bumped one in case of lenghty confirmation delay
 #[derive(Clone, PartialEq)]
-enum InputMaterial {
+pub(crate) enum InputMaterial {
 	Revoked {
 		script: Script,
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Not a new issue but can you add a commit renaming this to redeemscript or witness_redeemscript or whatever?

File: lightning/src/ln/channelmonitor.rs [hunk 16]
```diff
@@ -1630,127 +1461,40 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
...
 						}
+						let witness_data = InputMaterial::Revoked { script: expected_script, pubkey: Some(revocation_pubkey), key: revocation_key, is_htlc: true, revoked_amount: tx.output[transaction_output_index as usize].value };
+						outpoints.push(ClaimRequest { absolute_timelock: htlc.cltv_expiry, aggregable: true, outpoint: BitcoinOutPoint { txid: commitment_txid, vout: transaction_output_index }, witness_data });
 					}
 				}
 			}
 
-			if !inputs.is_empty() || !txn_to_broadcast.is_empty() || per_commitment_option.is_some() { // ie we're confident this is actually ours
+			// Last, track onchain revoked commitment transaction and fail backward outgoing HTLCs as payment path is broken
+			if !claimable_outpoints.is_empty() || per_commitment_option.is_some() { // ie we're confident this is actually ours
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: I dont think its possible for us to have inserted into claimable_outpoints by this point, only outpoints. I guess there are no tests which hit this?

File: lightning/src/ln/channelmonitor.rs [hunk 17]
```diff
@@ -1630,127 +1461,40 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
...
 				// We're definitely a remote commitment transaction!
-				log_trace!(self, "Got broadcast of revoked remote commitment transaction, generating general spend tx with {} inputs and {} other txn to broadcast", inputs.len(), txn_to_broadcast.len());
+				log_trace!(self, "Got broadcast of revoked remote commitment transaction, going to generate general spend tx with {} inputs", claimable_outpoints.len());
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Same here - I think you want outpoints, not claimable_outpoints.

File: lightning/src/ln/channelmonitor.rs (hunk 16 above)
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: Hmmm is this check accurate even using outpoints, given per_commitment_option.is_some (and we shouldn't care about tracking revoked remote commitment tx on chain if there is no htlc outputs, which may explain why we don't hit with a test)

File: lightning/src/ln/channelmonitor.rs (hunk 15 above)
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: Done

File: lightning/src/ln/onchaintx.rs [hunk 18]
```diff
@@ -0,0 +1,713 @@
...
+					return None;
+				}
+			}
+		}
+		log_trace!(self, "...with timer {}", new_timer);
+		assert!(predicted_weight >= bumped_tx.get_weight());
+		Some((new_timer, new_feerate, bumped_tx))
+	}
+
+	pub(super) fn block_connected<B: Deref, F: Deref>(&mut self, txn_matched: &[&Transaction], claim_requests: HashMap<Sha256dHash, Vec<ClaimRequest>>, height: u32, broadcaster: B, fee_estimator: F) -> Vec<SpendableOutputDescriptor>
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: claim_requests seems pretty redundant as-is - its a map from previous-txid to a vec of a struct which contains the previous-txid in the outpoint. Can we drop the duplication to make it a bit easier to see correctness?

File: lightning/src/ln/onchaintx.rs [hunk 19]
```diff
@@ -0,0 +1,713 @@
...
+			if let Some((new_timer, new_feerate, tx)) = self.generate_claim_tx(height, &claim_material, &*fee_estimator) {
+				claim_material.height_timer = new_timer;
+				claim_material.feerate_previous = new_feerate;
+				let txid = tx.txid();
+				new_pending_claim_requests.insert(txid, claim_material);
+				for k in claim.1.keys() {
+					new_claimable_outpoints.insert(k.clone(), (txid, height));
+				}
+				log_trace!(self, "Broadcast onchain {}", log_tx!(tx));
+				spendable_outputs.push(SpendableOutputDescriptor::StaticOutput {
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: We shouldn't generate this until the tx is confirmed.

File: lightning/src/ln/onchaintx.rs [hunk 20]
```diff
@@ -0,0 +1,713 @@
...
+	}
+
+	pub(super) fn block_connected<B: Deref, F: Deref>(&mut self, txn_matched: &[&Transaction], claim_requests: HashMap<Sha256dHash, Vec<ClaimRequest>>, height: u32, broadcaster: B, fee_estimator: F) -> Vec<SpendableOutputDescriptor>
+		where B::Target: BroadcasterInterface,
+		      F::Target: FeeEstimator
+	{
+		let mut new_claims = Vec::new();
+		let mut aggregated_claim = HashMap::new();
+		let mut aggregated_soonest = ::std::u32::MAX;
+		let mut new_pending_claim_requests = HashMap::with_capacity(new_claims.len());
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: new_claims.len() is always 0.

File: lightning/src/ln/onchaintx.rs [hunk 21]
```diff
@@ -0,0 +1,713 @@
...
+		let mut new_claimable_outpoints = HashMap::with_capacity(claim_requests.len());
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: Same here with claim_requests.len().

File: lightning/src/ln/onchaintx.rs [hunk 22]
```diff
@@ -0,0 +1,713 @@
...
+				});
+				broadcaster.broadcast_transaction(&tx);
+			}
+		}
+		for (k,v) in new_claimable_outpoints.drain() {
+			log_trace!(self, "Registering claiming request for {}:{}", k.txid, k.vout);
+			self.claimable_outpoints.insert(k, v);
+		}
+		for (k, v) in new_pending_claim_requests.drain() {
+			self.pending_claim_requests.insert(k, v);
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: Why not do this in the loop instead of having another map to track them before adding them later?

File: lightning/src/ln/onchaintx.rs [hunk 23]
```diff
@@ -0,0 +1,713 @@
...
+				spendable_outputs.push(SpendableOutputDescriptor::StaticOutput {
+					outpoint: BitcoinOutPoint { txid: tx.txid(), vout: 0 },
+					output: tx.output[0].clone(),
+				});
+				broadcaster.broadcast_transaction(&tx);
+			}
+		}
+		for (k,v) in new_claimable_outpoints.drain() {
+			log_trace!(self, "Registering claiming request for {}:{}", k.txid, k.vout);
+			self.claimable_outpoints.insert(k, v);
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: Why not do this in the loop instead of having another map to track them before adding them later?

File: lightning/src/ln/channelmonitor.rs [hunk 24]
```diff
@@ -445,72 +445,63 @@ struct LocalSignedTx {
...
-enum InputMaterial {
+pub(crate) enum InputMaterial {
 	Revoked {
-		script: Script,
+		witness_script: Script,
 		pubkey: Option<PublicKey>,
 		key: SecretKey,
 		is_htlc: bool,
-		amount: u64,
+		revoked_amount: u64,
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: I really dont understand the need for this commit - amount, revoked_amount, local_amount and remote_amount are all the amount as it appears in the transaction output. I think the new names may be even *more* confusing as it'll make me do a double-take to make sure we don't need to calculate them from some Channel data instead of from on-chain outputs.

File: lightning/src/ln/onchaintx.rs [hunk 25]
```diff
@@ -0,0 +1,713 @@
...
+
+		// Compute new height timer to decide when we need to regenerate a new bumped version of the claim tx (if we
+		// didn't receive confirmation of it before, or not enough reorg-safe depth on top of it).
+		let new_timer = Self::get_height_timer(height, cached_claim_datas.soonest_timelock);
+		let mut inputs_witnesses_weight = 0;
+		let mut total_amount = 0;
+		for per_outp_material in cached_claim_datas.per_input_material.values() {
+			match per_outp_material {
+				&InputMaterial::Revoked { ref witness_script, ref is_htlc, ref revoked_amount, .. } => {
+					inputs_witnesses_weight += Self::get_witnesses_weight(if !is_htlc { &[InputDescriptors::RevokedOutput] } else if HTLCType::scriptlen_to_htlctype(witness_script.len()) == Some(HTLCType::OfferedHTLC) { &[InputDescriptors::RevokedOfferedHTLC] } else if HTLCType::scriptlen_to_htlctype(witness_script.len()) == Some(HTLCType::AcceptedHTLC) { &[InputDescriptors::RevokedReceivedHTLC] } else { &[] });
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Done

File: lightning/src/ln/onchaintx.rs (hunk 18 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Yeah, my initial worry was someone inadvertently introducing cross-block aggregation, segmenting by txid was the reason.

File: lightning/src/ln/onchaintx.rs (hunk 19 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Yes but that's already current behavior, pre-refactoring. Spendable output generation after ANTI_REORG_SAFE_DELAY is a todo in another PR

File: lightning/src/ln/onchaintx.rs (hunk 22 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: I thought borrow-checker would complain for some ownership already taken on `self`, but it works well, `generate_claim_tx` isn't mutable (it used to be in a previous verison, or at least I tried with one...)

File: lightning/src/ln/onchaintx.rs (hunk 18 above)
Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs]: Right, but why not just have it be a map from previous-txid to something that contains the vout, instead of a full OutPoint?

File: lightning/src/ln/onchaintx.rs (hunk 19 above)
Comment (TheBlueMatt) [reply lightning/src/ln/onchaintx.rs]: Grr that sucks. Alright, we should improve that, then.

File: lightning/src/ln/channelmonitor.rs (hunk 24 above)
Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: Okay was in reaction to https://github.com/rust-bitcoin/rust-lightning/pull/462#discussion_r379195303, I'm ~0 on this, dropped the commit

File: lightning/src/ln/onchaintx.rs [hunk 26]
```diff
@@ -0,0 +1,705 @@
...
+	pub(super) fn block_connected<B: Deref, F: Deref>(&mut self, txn_matched: &[&Transaction], claimable_outpoints: Vec<Vec<ClaimRequest>>, height: u32, broadcaster: B, fee_estimator: F) -> Vec<SpendableOutputDescriptor>
+		where B::Target: BroadcasterInterface,
+		      F::Target: FeeEstimator
+	{
+		let mut new_claims = Vec::new();
+		let mut aggregated_claim = HashMap::new();
+		let mut aggregated_soonest = ::std::u32::MAX;
+		let mut spendable_outputs = Vec::new();
+
+		// Try to aggregate outputs if they're 1) belong to same parent tx, 2) their
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: This comment is wrong now - we may aggregate an HTLC claim which spent a commitment tx with a spend of the commitment tx. This is, of course, fine, as any observer doesn't learn anything from that behavior, but the comment should be fixed.

File: lightning/src/ln/onchaintx.rs [hunk 27]
```diff
@@ -0,0 +1,705 @@
...
+					return None;
+				}
+			}
+		}
+		log_trace!(self, "...with timer {}", new_timer);
+		assert!(predicted_weight >= bumped_tx.get_weight());
+		Some((new_timer, new_feerate, bumped_tx))
+	}
+
+	pub(super) fn block_connected<B: Deref, F: Deref>(&mut self, txn_matched: &[&Transaction], claimable_outpoints: Vec<Vec<ClaimRequest>>, height: u32, broadcaster: B, fee_estimator: F) -> Vec<SpendableOutputDescriptor>
```
Comment (TheBlueMatt) [inline lightning/src/ln/onchaintx.rs]: Please drop the Vec<Vec<>>. That seems needlessly inefficient.

**Comment by TheBlueMatt** [comment]
I'm gonna have a few follow-ups anyway that are only partially-related, so I'll just take the above comments with those.

File: lightning/src/ln/onchaintx.rs (hunk 26 above)
Comment (ariard) [reply lightning/src/ln/onchaintx.rs]: Hmmm why comment is wrong ?  Spending a HTLC output and to_local/to_remote output that still spending same commitment tx, sorry your comment is unclear to me.
